homeassistant/components/isy994/binary_sensor.py

```diff
@@ -28,83 +65,187 @@ def setup_platform(
     child_nodes = []
 
     for node in hass.data[ISY994_NODES][BINARY_SENSOR]:
-        if node.parent_node is None:
-            device = ISYBinarySensorEntity(node)
-            devices.append(device)
-            devices_by_address[node.address] = device
+        device_class, device_type = _detect_device_type_and_class(node)
+        if node.protocol == PROTO_INSTEON:
+            if node.parent_node is not None:
+                # We'll process the Insteon child nodes last, to ensure all parent
+                # nodes have been processed
+                child_nodes.append((node, device_class, device_type))
+                continue
+            device = ISYInsteonBinarySensorEntity(node, device_class)
         else:
-            # We'll process the child nodes last, to ensure all parent nodes
-            # have been processed
-            child_nodes.append(node)
-
-    for node in child_nodes:
-        try:
-            parent_device = devices_by_address[node.parent_node.address]
-        except KeyError:
-            _LOGGER.error(
-                "Node %s has a parent node %s, but no device "
-                "was created for the parent. Skipping.",
-                node.address,
-                node.primary_node,
-            )
-        else:
-            device_type = _detect_device_type(node)
-            subnode_id = int(node.address[-1], 16)
-            if device_type in ("opening", "moisture"):
-                # These sensors use an optional "negative" subnode 2 to snag
-                # all state changes
-                if subnode_id == 2:
-                    parent_device.add_negative_node(node)
-                elif subnode_id == 4:
-                    # Subnode 4 is the heartbeat node, which we will represent
-                    # as a separate binary_sensor
-                    device = ISYBinarySensorHeartbeat(node, parent_device)
-                    parent_device.add_heartbeat_device(device)
-                    devices.append(device)
-            else:
-                # We don't yet have any special logic for other sensor types,
-                # so add the nodes as individual devices
-                device = ISYBinarySensorEntity(node)
+            device = ISYBinarySensorEntity(node, device_class)
+        devices.append(device)
+        devices_by_address[node.address] = device
+
+    # Handle some special child node cases for Insteon Devices
+    for (node, device_class, device_type) in child_nodes:
```

Comment:
Nice job on the cleanup here.  Thank you for replacing the magic strings.  This so much easier to understand what is going on.